src/Servers/Kestrel/Core/src/Internal/Http2/Http2Stream.cs

```diff
@@ -198,6 +201,15 @@ protected override bool TryParseRequest(ReadResult result, out bool endConnectio
         // do the reading from a pipeline, nor do we use endConnection to report connection-level errors.
         endConnection = !TryValidatePseudoHeaders();
 
+        // 431 if the headers are too large
+        if (TotalParsedHeaderSize > _context.ServiceContext.ServerOptions.Limits.MaxRequestHeadersTotalSize)
+        {
+            KestrelBadHttpRequestException.Throw(RequestRejectionReason.HeadersExceedMaxTotalSize);
+        }
+
+        // 431 if we received too many headers
+        CheckRequestHeadersCountLimit(final: true);
```

Comment:
```suggestion
        if (RequestHeadersParsed > ServerOptions.Limits.MaxRequestHeaderCount * 2)
        {
            KestrelBadHttpRequestException.Throw(RequestRejectionReason.TooManyHeaders);
        }
```

This is simple enough to inline, and I think we should get rid of `CheckRequestHeadersCountLimit`.